closed/src/java.base/share/classes/sun/security/ec/NativeECDHKeyAgreement.java

```diff
@@ -116,13 +116,19 @@ private void init(Key key)
         /* attempt to translate the key if it is not an ECKey */
         ECKey ecKey = ECKeyFactory.toECKey(key);
         if (ecKey instanceof ECPrivateKey ecPrivateKey) {
-            Optional<ECOperations> opsOpt =
-                ECOperations.forParameters(ecPrivateKey.getParams());
-            if (opsOpt.isEmpty()) {
-                NamedCurve nc = CurveDB.lookup(ecPrivateKey.getParams());
-                throw new InvalidAlgorithmParameterException(
+            Optional<ECOperations> opsOpt = null;
+            try {
+                opsOpt = ECOperations.forParameters(ecPrivateKey.getParams());
+                if (opsOpt.isEmpty()) {
+                    NamedCurve nc = CurveDB.lookup(ecPrivateKey.getParams());
+                    throw new InvalidAlgorithmParameterException(
                         "Curve not supported: " +
                         ((nc != null) ? nc.toString() : "unknown"));
+                }
+            } catch (InvalidAlgorithmParameterException iape) {
+                if (!(ECUtil.isBrainpoolP512r1(ecPrivateKey.getParams()))) {
+                    throw iape;
+                }
             }
```

Comment:
Suggest moving the declaration of `params` earlier so it can be used here, and that we should avoid throwing exceptions only to ignore them:
```java
            ECParameterSpec params = ecPrivateKey.getParams();
            ECOperations ops = ECOperations.forParameters(params).orElse(null);
            if ((ops == null) && !NativeECUtil.isBrainpoolP512r1(params)) {
                NamedCurve nc = CurveDB.lookup(params);
                throw new InvalidAlgorithmParameterException(
                        "Curve not supported: " +
                        ((nc != null) ? nc.toString() : "unknown"));
            }
```

Reply:
Updated as suggested to avoid using exception handling for flow control. The declaration of the `params` variable is now just prior to this code block.